### PR TITLE
[Sparse] Add RGCN example

### LINKED ISSUE #5375: [Sparse] Add a heterogeneous graph example RGCN.
## 🔨Work Item

**IMPORTANT:**
* This template is only for dev team to track project progress. For feature request or bug report, please use the corresponding issue templates.
* DO NOT create a new work item if the purpose is to fix an existing issue or feature request. We will directly use the issue in the project tracker.

Project tracker: https://github.com/orgs/dmlc/projects/2

### PR DESCRIPTION
**Pull request by czkkkkkk**
## Description
Resolve #5375 .

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/sparse/rgcn.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
+"""Modeling Relational Data with Graph Convolutional Networks
+Paper: https://arxiv.org/abs/1703.06103
+Reference Code: https://github.com/tkipf/relational-gcn
+"""
```
Comment (frozenbugs) [inline examples/sparse/rgcn.py:4]: Follow existing convention:
https://github.com/dmlc/dgl/blob/master/examples/sparse/sign.py
```
"""
[Modeling Relational Data with Graph Convolutional Networks]
(https://arxiv.org/abs/1703.06103)

Reference Code: https://github.com/tkipf/relational-gcn
"""
```

File: examples/sparse/rgcn.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+import argparse
+
+import dgl.sparse as dglsp
+import torch as th
```
Comment (frozenbugs) [inline examples/sparse/rgcn.py:8]: just `import torch`

Comment (frozenbugs) [reply examples/sparse/rgcn.py:8]: https://github.com/dmlc/dgl/tree/master/examples/sparse

File: examples/sparse/rgcn.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+import torch.nn as nn
+import torch.nn.functional as F
+
+from dgl.data.rdf import AIFBDataset, AMDataset, BGSDataset, MUTAGDataset
+
+
+class RelGraphConvLayer(nn.Module):
+    def __init__(self, in_feat, out_feat, rel_names):
```
Comment (frozenbugs) [inline examples/sparse/rgcn.py:16]: in_size, out_size?

**Comment by frozenbugs** [comment]
Can you familiar with our existing examples and follow the conventions?

File: examples/sparse/rgcn.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        super(RelGraphConvLayer, self).__init__()
+        self.in_feat = in_feat
+        self.out_feat = out_feat
+        self.rel_names = rel_names
```
Comment (frozenbugs) [inline examples/sparse/rgcn.py:20]: what's rel_names?

File: examples/sparse/rgcn.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+    def create_adjs(g):
+        adjs = {}
+        for rel in g.canonical_etypes:
+            stype, _, dtype = rel
+            row, col = g.edges(etype=rel)
+            adjs[rel] = dglsp.from_coo(
+                col,
+                row,
+                shape=(g.number_of_nodes(dtype), g.number_of_nodes(stype)),
```
Comment (frozenbugs) [inline examples/sparse/rgcn.py:97]: num_nodes

File: examples/sparse/rgcn.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    test_idx = th.nonzero(test_mask, as_tuple=False).squeeze()
+    labels = g.nodes[category].data.pop("labels")
+    val_idx = train_idx
+
+    def create_adjs(g):
+        adjs = {}
+        for rel in g.canonical_etypes:
+            stype, _, dtype = rel
+            row, col = g.edges(etype=rel)
+            adjs[rel] = dglsp.from_coo(
```
Comment (frozenbugs) [inline examples/sparse/rgcn.py:94]: spmatrix?

**Comment by frozenbugs** [comment]
Also please add test case here.
tests/examples/test_sparse_examples.py

**Comment by jermainewang** [comment]
I'm thinking whether we indeed need this example because it generally demonstrates the usage of `HeteroGraphConv` same as https://github.com/dmlc/dgl/blob/master/examples/pytorch/rgcn-hetero/model.py#L61 . Considering we will have a core example like that, what else does this example focus on?

File: examples/sparse/rgcn.py (hunk 4 above)
Comment (czkkkkkk) [reply examples/sparse/rgcn.py]: Relation name (name of edge type).

**Comment by czkkkkkk** [comment]
> Can you familiar with our existing examples and follow the conventions?

Refined this example following `sgc.py`

**Comment by czkkkkkk** [comment]
> I'm thinking whether we indeed need this example because it generally follows the idea of HeteroGraphConv same as https://github.com/dmlc/dgl/blob/master/examples/pytorch/rgcn-hetero/model.py#L61 . Considering we will have a core example like that, what else does this example focus on?

I think this example simply demonstrates how sparse APIs handle heterogeneous graphs. We may want to show there is no functionality loss when migrating from heterogeneous graph to sparse matrix.

**Comment by jermainewang** [comment]
My point is we probably should show an example using `HeteroGraphConv` which is the recommended way of dealing with heterogeneous graphs.
